### PR TITLE
[#IOPID-495] add spid level management

### PR DESCRIPTION
**Pull request by arcogabbo**
#### List of Changes

* added spid level management
* added `spid_level` required field in `CommonTokenUser`
* decoupled acs and token handlers from `configOrThrow` to facilitate unit testing
* added `acs` handler happy path unit test
* removed some unnecessary lines
* added some spid utilities under `src/utils/spid.ts`

#### Motivation and Context

add spid level field in JWT

#### How Has This Been Tested?

unit and integration tests through docker

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/utils/spid.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { SAML_NAMESPACE } from "@pagopa/io-spid-commons/dist/utils/saml";
+import { enumType } from "@pagopa/ts-commons/lib/types";
+import { flow } from "fp-ts/lib/function";
+import * as O from "fp-ts/lib/Option";
+import * as t from "io-ts";
+
+export enum SpidLevelEnum {
```
Comment (gquadrati) [inline src/utils/spid.ts:7]: In future, we should add it within fn-commons definitions.yaml.

**Review by gquadrati — DISMISSED** [review]
LGTM 🚀

File: src/handlers/spid.ts [hunk 2]
```diff
@@ -117,3 +141,159 @@ export const accessLogHandler = (
...
+            assertion,
+            safeXMLParseFromString,
+            O.chain(getSpidLevelFromSAMLResponse),
+            E.fromOption(() => new Error("Spid level retrieval failed"))
+          )
+        ),
+        E.getOrElse(() => SpidLevelEnum["https://www.spid.gov.it/SpidL2"]),
+        E.of
+      )
+    ),
```
Comment (BurnedMarshal) [inline src/handlers/spid.ts:169]: Why we are parsing the `SAMLResponse`, the value we need is already parsed from `io-spid-commons` and included into the property `authnContextClassRef` of the user.
Just edit the `SpidUser` type adding the missing property.
```suggestion
```

File: src/types/user.ts [hunk 3]
```diff
@@ -31,6 +31,15 @@ export const SpidUser = t.intersection([
...
 export type SpidUser = t.TypeOf<typeof SpidUser>;
 
+export const SpidUserWithLevel = t.intersection([
+  SpidUser,
+  t.type({
+    authnContextClassRef: SpidLevel
+  })
+]);
+
+export type SpidUserWithLevel = t.TypeOf<typeof SpidUserWithLevel>;
```
Comment (BurnedMarshal) [inline src/types/user.ts:41]: Can we extend the `SpidUser`?

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: src/handlers/spid.ts (hunk 2 above)
Comment (gquadrati) [reply src/handlers/spid.ts:169]: There is no `authnContextClassRef` within the user coming from `spid-commons`. 
In io-backend we also retrieve it from SAML Response (see [here](https://github.com/pagopa/io-backend/blob/f781120073e09a360043e2d78da712497c2d0522/src/types/user.ts#L253-L254)).

It's true that we should refactor spid-commons to return it, but we can postpone to another PR.

Comment (BurnedMarshal) [reply src/handlers/spid.ts:169]: My bad, the value is not returned from `io-spid-commons`, we need to extract it from the SAMLResponse

File: src/handlers/spid.ts [hunk 4]
```diff
@@ -117,3 +141,159 @@ export const accessLogHandler = (
     }
   );
 };
+
+export const acs: (
```
[the remark this answers is not in the document]
Comment (gquadrati) [reply src/handlers/spid.ts]: Done in [small refactor + L1 as default spid level](https://github.com/pagopa/hub-spid-login-ms/pull/90/commits/c68fbf3877305e2889e21a46b9b871e841ce1df2)

**Review by BurnedMarshal — APPROVED** [review]
LGTM

**Pull request merged**